toc/rfc/rfc-draft-generic-per-route-features.md

```diff
@@ -0,0 +1,90 @@
+# Meta
+[meta]: #meta
+- Name: Generic Per-Route Features
+- Start Date: 2024-01-08
+- Author(s): @maxmoehl, @domdom82
+- Status: Draft <!-- Acceptable values: Draft, Approved, On Hold, Superseded -->
+- RFC Pull Request: (fill in with PR link after you submit it)
+
+## Summary
+
+## Problem
+
+Implementing per-route features in Cloud Foundry is a tedious process because it requires
+changes to multiple components in multiple working groups. Most components will only need to be
+made aware of new fields without adding any significant logic but the over-head of making the
+change still exists.
+
+On the other side are the components that actually need to be aware of the change. Mostly the
+routing components (gorouter, maybe route-emitter / -registrar), the API (cloud controller)
+and UIs (CLI).
+
+## Proposal
+
+To reduce the overall overhead of implementing new per-route features we propose to add a generic
+key-value map to the definition of a route across all components. All components handling the map
+SHOULD ignore the contents of the map except gorouter, cloud controller and cf CLI which MUST handle
+the map and its values accordingly. Other components MAY limit the size of the map or size of
+keys / values for technical reasons.
+
+### Required Changes
+
+#### CLI
+
+Support for a generic map MUST be added to the manifest. There MAY be either individual command line
+flags or a single flag which supports generic key-value pairs for the `map-route` sub-command.
+
+#### Cloud Controller
+
+The API MUST specify and implement a new field to carry the map inside the route object. The API
+MAY implement per-field validations as features of the map are specified. The route MAY be
+rejected with an appropriate error message if the provided map is invalid and MUST be passed on
+as-is otherwise. The API also MUST store the map as a generic key-value map to ensure changes to 
+the map do not require a change to the database schema.
+
+To-Do:
+* Make the validation required?
+
+#### BBS, Rep, Executor, Route-Emitter
+
+The diego components MUST add support for the new field specified by the API and forward the
+contents as-is without modification or, if the field exceeds technical limitations, MUST raise an 
+error.
+
+#### Gorouter
+
+Gorouter MUST accept the new field and parse its contents. If the map cannot be parsed it MUST 
+be ignored. If individual key-value pairs are invalid those MUST be ignored and other key-value 
+pairs MUST still be parsed and acted upon.
+
+It MAY be decided to implement one feature as part of this RFC which would be implemented together
+with the generic logic for supporting the new map.
+
+As new features are specified they MUST be implemented in gorouter.
```

Comment:
```suggestion
As new features are specified they MUST be implemented in gorouter.

**Example router.register NATS message containing additional options**

```
{
  "host": "127.0.0.1",
  "port": 4567,
  "tls_port": 1234,
  "protocol": "http1",
  "uris": [
    "my_first_url.localhost.routing.cf-app.com",
    "my_second_url.localhost.routing.cf-app.com"
  ],
  "tags": {
    "another_key": "another_value",
    "some_key": "some_value"
  },
  "options": {
    "lb-algo": "least-conn",
    "conn-limit": 100,
    "cookie": "my-session-cookie" 
  },
  "app": "some_app_guid",
  "stale_threshold_in_seconds": 120,
  "private_instance_id": "some_app_instance_id",
  "isolation_segment": "some_iso_seg_name",
  "server_cert_domain_san": "some_subject_alternative_name"
}
```
```

Reply:
provided an example NATS message that contains all data types we would want to support (excluding floats).

Reply:
There are only [numbers](https://datatracker.ietf.org/doc/html/rfc8259#section-6) in JSON :)